src/main/java/com/breskul/bibernate/persistence/PersistenceContext.java

```diff
@@ -10,8 +15,9 @@
 
 public class PersistenceContext {
 
-  private final Map<EntityKey<?>, Object> firstLevelCache = new ConcurrentHashMap<>(); //TODO Think what better HashMap vs ConcurrentHashMap
-  private final Map<EntityKey<?>, Object[]> entitySnapshots = new ConcurrentHashMap<>(); // for dirty checking
+  private final Map<EntityKey<?>, Object> firstLevelCache = new ConcurrentHashMap<>();
+  private final Map<EntityKey<?>, List<Pair<String, Object>>> entitySnapshots = new ConcurrentHashMap<>();
+  private final Map<EntityKey<?>, List<Triple<? extends Class<?>, String, Object>>> toOneRelationSnapshots = new ConcurrentHashMap<>();
```

Comment:
Instead of `Pair` and `Triple` I suggest to use here concrete classes or records, because current implementation is really hard to read.

Reply:
done